extensions/wikia/Email/EmailController.class.php

```diff
@@ -0,0 +1,292 @@
+	 * method is used for testing only, to preview the email that will be sent.
+	 *
+	 * @template Email_preview
+	 *
+	 * @throws \MWException
+	 */
+	public function handle() {
+		// If something previously has thrown an error (likely 'init') don't continue
+		if ( $this->hasErrorResponse ) {
+			return;
+		}
+
+		try {
+			$this->assertCanEmail();
+
+			$toAddress = $this->getToAddress();
+			$fromAddress = $this->getFromAddress();
+			$replyToAddress = $this->getReplyToAddress();
+			$subject = $this->getSubject();
+			$body = [ 'html' => $this->getBody() ];
+
+			if ( !$this->test ) {
+				$status = \UserMailer::send(
+					$toAddress,
+					$fromAddress,
+					$subject,
+					$body,
+					$replyToAddress
+				);
+				$this->assertGoodStatus( $status );
+			}
+		} catch ( ControllerException $e ) {
+			$this->setErrorResponse( $e );
+			return;
+		}
+
+		$this->response->setData( [
+			'result' => 'ok',
+			'subject' => $subject,
+			'body' => $body['html'],
+		] );
+
+		return;
+	}
+
+	/**
+	 * Create an error response for any exception thrown while creating this email
+	 *
+	 * @param ControllerException $e
+	 *
+	 */
+	protected function setErrorResponse( ControllerException $e ) {
+		$this->hasErrorResponse = true;
+		$this->response->setData( [
+			'result' => $e->errorType,
+			'msg' => $e->getMessage(),
+		] );
+		return;
+	}
+
+	protected function getTargetLang() {
+		return $this->targetUser->getOption( 'language' );
+	}
+
+	/**
+	 * Returns the address for the recipient of this email
+	 *
+	 * @return \MailAddress
+	 */
+	protected function getToAddress() {
+		$to = new \MailAddress( $this->targetUser->getEmail(), $this->targetUser->getName() );
+		return $to;
+	}
+
+	/**
+	 * Returns the address for the sender of this email
+	 *
+	 * @return \MailAddress
+	 */
+	protected function getFromAddress() {
+		$sender = new \MailAddress( $this->wg->PasswordSender, $this->wg->PasswordSenderName );
+		return $sender;
+	}
+
+	/**
+	 * Get the 'reply-to' address for the send of this email, if different than the 'from' address.
+	 *
+	 * @return \MailAddress|null
+	 */
+	protected function getReplyToAddress() {
+		$replyAddr = null;
+		if ( !empty( $this->wg->NoReplyAddress ) ) {
+			$name = wfMessage( 'emailext-no-reply-name' )->escaped();
+			$replyAddr = new \MailAddress( $this->wg->NoReplyAddress, $name );
+		}
+
+		return $replyAddr;
+	}
+
+	/**
+	 * Return the subject used for this email
+	 */
+	abstract protected function getSubject();
+
+	/**
+	 * Renders the 'body' view of the current email controller
+	 *
+	 * @return string
+	 */
+	protected function getBody() {
+		$html = $this->app->renderView(
+			get_class( $this ),
+			'body',
+			$this->request->getParams()
+		);
+		return $html;
+	}
+
+	protected function findUserFromRequest( $paramName, \User $default = null ) {
+		$userName = $this->getRequest()->getVal( $paramName );
+		if ( empty( $userName ) ) {
+			return $default;
```

Comment:
Should we make sure that `$default` is a user object before returning it? It might be nice to check and throw an exception like we do for `$paramName` since `$default` could be null. Or, we could make the default argument `$this->wg->User`. That could work as well.